[FX] Normalize torch. arithmetic operators to python operator.*

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54236 [FX] Normalize torch. arithmetic operators to python operator.***

Add a pass to normalize operator targets. Currently it resolves `operator.*` ops to the correspoinding `torch` ops when we can statically reason that the operator would dispatch to a torch op

Differential Revision: [D27149411](https://our.internmc.facebook.com/intern/diff/D27149411)

Commit:
[FX] Normalize Python operators to `torch.` ops when called with Tensors

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize Python operators to `torch.` ops when appropriate"


Differential Revision: [D27149411](https://our.internmc.facebook.com/intern/diff/D27149411)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize Python operators to `torch.` ops when appropriate"


Differential Revision: [D27149411](https://our.internmc.facebook.com/intern/diff/D27149411)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize Python operators to `torch.` ops when appropriate"


Differential Revision: [D27149411](https://our.internmc.facebook.com/intern/diff/D27149411)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize Python operators to `torch.` ops when appropriate"


Differential Revision: [D27149411](https://our.internmc.facebook.com/intern/diff/D27149411)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize Python operators to `torch.` ops when appropriate"


Differential Revision: [D27149411](https://our.internmc.facebook.com/intern/diff/D27149411)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize Python operators to `torch.` ops when appropriate"


Add a pass to normalize operator targets. Currently it resolves `operator.*` ops to the correspoinding `torch` ops when we can statically reason that the operator would dispatch to a torch op

Differential Revision: [D27149411](https://our.internmc.facebook.com/intern/diff/D27149411)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize Python operators to `torch.` ops when appropriate"


Add a pass to normalize operator targets. Currently it resolves `operator.*` ops to the correspoinding `torch` ops when we can statically reason that the operator would dispatch to a torch op

Differential Revision: [D27149411](https://our.internmc.facebook.com/intern/diff/D27149411)

[ghstack-poisoned]

Commit:
Update on "[FX] Normalize torch. arithmetic operators to python operator.*"


Add a pass to normalize operator targets. Currently it resolves `operator.*` ops to the correspoinding `torch` ops when we can statically reason that the operator would dispatch to a torch op

Differential Revision: [D27149411](https://our.internmc.facebook.com/intern/diff/D27149411)

[ghstack-poisoned]

zdevito commented on torch/fx/experimental/normalize.py:
```diff
@@ -117,3 +120,52 @@ def _args_kwargs_to_normalized_kwargs(self, sig : inspect.Signature, args : Tupl
+        torch.sub : operator.sub,
+        torch.div : operator.truediv,
+        torch.floor_divide : operator.floordiv,
+        torch.remainder : operator.mod,
```
double check % does the same thing, since they are not exactly the same operator (requires testing with negative numbers) (https://rob.conery.io/2018/08/21/mod-and-remainder-are-not-the-same/)

jamesr66a replied:
Would the argument for the `operator.mod -> torch.remainder` direction still apply here? This is going to get dispatched down to `torch.remainder` judging by what `python_variable_methods` does so it it consistent with PT?

zdevito commented on torch/fx/experimental/schema_type_annotation.py:
```diff
@@ -33,6 +33,24 @@ def __init__(self, module : torch.nn.Module, annotate_functionals : bool = True,
+            self._cur_node = n
+            return super().run_node(n)
+        finally:
+            self._cur_node = None
```
this seems a bit messy, to it is more related to the abstraction for how transforms are done and not this PR in particular.

zdevito replied:
Is there a reason this strives to preserve type for param nodes but nothing else?

jamesr66a replied:
Types were set for everything else in a previous PR, it was just missing for `placeholders`. This adds the missing annotations. Though this change is actually no longer needed now that we don't rely on types, let me just remove it

zdevito reviewed (approved):
Seems good now, some comments unrelated to this particular change

jamesr66a commented on torch/fx/experimental/normalize.py:
```diff
@@ -117,3 +120,87 @@ def _args_kwargs_to_normalized_kwargs(self, sig : inspect.Signature, args : Tupl
+        operator.sub : torch.sub,
+        operator.truediv : torch.div,
+        operator.floordiv: torch.floor_divide,
+        operator.mod: torch.remainder,
```
Going in the `operator.mod` -> `torch.remainder` direction comes directly from the mapping in `python_variable_methods`: https://github.com/pytorch/pytorch/blob/fd58ececab731803eceace70b985b35f9872b53c/tools/autograd/templates/python_variable_methods.cpp#L1150 so I assume this is correct unless PyTorch has it wrong?

Commit:
Update on "[FX] Normalize torch. arithmetic operators to python operator.*"


Add a pass to normalize operator targets. Currently it resolves `operator.*` ops to the correspoinding `torch` ops when we can statically reason that the operator would dispatch to a torch op

Differential Revision: [D27149411](https://our.internmc.facebook.com/intern/diff/D27149411)

[ghstack-poisoned]
